source/linker.cpp

```diff
@@ -0,0 +1,1098 @@
+// Copyright (c) 2017 Pierre Moreau
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "spirv/1.2/spirv.hpp11"
+
+#include <cassert>
+#include <cstring>
+
+#include <algorithm>
+#include <functional>
+#include <vector>
+#include <unordered_map>
+#include <unordered_set>
+
+#include "diagnostic.h"
+#include "table.h"
+
+
+namespace {
+
+using spirv_word = uint32_t;
+
+spirv_word GetWord(const uint32_t* binary, size_t word_offset) {
+  return binary[word_offset];
+}
+
+void SetWord(uint32_t* binary, size_t word_offset, spirv_word word) {
+  binary[word_offset] = word;
+}
+
+const char* GetString(const uint32_t* binary, size_t word_offset) {
+  return reinterpret_cast<const char*>(binary + word_offset);
+}
+
+/// Extracts the opcode and the number of words making up this instruction.
```

Comment:
If you use the binary parser then this is decomposed for you. (see below for higher level feedback at TransformBinary)